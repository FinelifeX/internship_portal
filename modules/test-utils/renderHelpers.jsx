/* eslint-disable import/no-extraneous-dependencies */
import React from 'react';
import TR from 'react-test-renderer';
import { render as RTLRender } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

/**
 * @typedef {Object.<string, any>} Props
 */

/**
 * @typedef {React.ReactNode} Children
 */

/**
 * @typedef {{props?: Props, children?: Children}} RenderOptions
 */

/**
 *
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @returns {JSX.Element}
 */
export const render = (Component, options = {}) => {
  const { props, children } = options;
  return <Component {...props}>{children}</Component>;
};

/**
 * Render React component inside MemoryRouter
 * @param {React.ComponentType} Component
 * @param {RenderOptions} [options]
 * @param {import('react-router-dom').MemoryRouterProps} routerOptions
 * @returns {JSX.Element}
 */
export const renderWithRouter = (
  Component,
  options = {},
  routerOptions = {},
) => {
  const { props, children } = options;

  return (
    <MemoryRouter {...routerOptions}>
      <Component {...props}>{children}</Component>
    </MemoryRouter>
  );
};

/**
 * Shallow render component using "react-test-renderer"
 * @type {import('react-test-renderer').create}
 */
export const shallow = TR.create;

/**
 * Render component using "render()" from "@testing-library/react"
 * @type {import('@testing-library/react').render}
 */
export const mount = RTLRender;
