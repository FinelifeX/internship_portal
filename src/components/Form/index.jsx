import React from 'react';
import PT from 'prop-types';
import classNames from 'classnames';
import styles from './Form.module.css';

const Form = ({ children, className, onSubmit }) => (
  <form
    data-testid="form"
    className={classNames(styles.form, className)}
    onSubmit={onSubmit}
  >
    {children}
  </form>
);

Form.defaultProps = {
  className: '',
  onSubmit: () => {},
};

Form.propTypes = {
  children: PT.node.isRequired,
  className: PT.string,
  onSubmit: PT.func,
};

export default Form;
