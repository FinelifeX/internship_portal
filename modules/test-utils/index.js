import * as RenderHelpers from './renderHelpers';

export { RenderHelpers };
