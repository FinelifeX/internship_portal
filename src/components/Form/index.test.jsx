import { shallow, mount, render } from 'test-utils/renderHelpers';
import { fireEvent } from '@testing-library/react';
import Form from './index';

describe('Form component', () => {
  const TEST_ID_FORM = 'form';
  const defaultProps = {
    onSubmit: jest.fn(),
  };
  const renderComponent = (props) =>
    render(Form, { props: { ...defaultProps, ...props }, children: 'hello' });

  it('should render correctly', () => {
    const tree = shallow(renderComponent()).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('should call onSubmit 1 time when submit occurs', () => {
    const { getByTestId } = mount(renderComponent());

    fireEvent.submit(getByTestId(TEST_ID_FORM));

    expect(defaultProps.onSubmit).toBeCalledTimes(1);
  });
});
