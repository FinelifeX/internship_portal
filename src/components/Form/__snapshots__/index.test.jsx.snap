// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component should render correctly 1`] = `
<form
  className="form"
  data-testid="form"
  onSubmit={[MockFunction]}
>
  hello
</form>
`;
